### PR TITLE
Ensemble functions

### PR DESCRIPTION
**Pull request by Wesleylewis05**
created perform_ensemble_functions which executes all ensemble functions in a json file.

### CODE REVIEW

**Review by hgrehm — CHANGES_REQUESTED** [review]
- Instead of looping through all ensembles in `dummy_ensembles.json`, we should just take the current ensemble in `active_ensembles.json`. This file should already have the current ensemble noted, and you can index into the ensemble list at the current ensemble index to know which functions to perform on this call. 
- Please sort imports (abc order)
- Please update the README with the new json file format (now including module name/directory, class, and function)

**Comment by ntlhui** [comment]
Note that it does not make sense to execute a member function - only static or standalone functions should be executed, as the scheduler should not guarantee existence of any objects while executing a function.

If class context is required, then the entry point function should be responsible for bootstrapping the appropriate object.

File: ensemble_tests/tester_functions.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+class SimpleFunctions():
+    def hello_world():
```
Comment (ntlhui) [inline ensemble_tests/tester_functions.py]: member function signature should be `def hello_world(self)`

File: ensemble_tests/tester_functions.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+        print("Hello World")
+
+    def print_this(string):
```
Comment (ntlhui) [inline ensemble_tests/tester_functions.py]: member function signature should be `def print_this(self, string: str)`

File: ensemble_tests/tester_functions.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+        print(string)
+
+    def add_this(x, y):
```
Comment (ntlhui) [inline ensemble_tests/tester_functions.py]: member function signature should be `def add_this(self, x, y)`

File: ensemble_tests/tester_functions.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+        print(x + y)
+
+    def subtract_this(x, y):
```
Comment (ntlhui) [inline ensemble_tests/tester_functions.py]: member function signature should be `def subtract_this(self, x, y)`

File: ensemble_tests/tester_functions.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+        print(x - y)
+
+    def multiply_this(x, y):
```
Comment (ntlhui) [inline ensemble_tests/tester_functions.py]: member function signature should be `def multiply_this(self, x, y)`

File: dummy_ensembles.json [hunk 6]
```diff
@@ -14,8 +17,11 @@
     },
     {
       "title": "zeta2",
-      "function": "someFunc",
-      "inputs": [],
+      "module_name": "tester_functions",
+      "module_directory": "ensemble_tests/tester_functions.py",
+      "class": "SimpleFunctions",
+      "function": "print_this",
```
Comment (ntlhui) [inline dummy_ensembles.json]: Consider is this necessarily the appropriate way to define this?

could we potentially define this as `ensemble_tests.tester_functions.SimpleFunctions.print_this`

**Review by ntlhui — CHANGES_REQUESTED** [review]

File: README.md [hunk 7]
```diff
@@ -28,29 +28,49 @@ Functions to perform should be stored in the following json format:
...
+                2
+            ],
+            "start_time": {
+                "hour": 12,
+                "minute": 0,
+                "second": 0
+            },
+            "iterations": 5,
+            "interval": 60
+        }
```
Comment (hgrehm) [inline README.md]: Could we make these examples in the README more generic? For example, the first one might have `"function": "path_to_function:print_function"` to illustrate that we have a path and a function name, separated by a dot.

File: README.md [hunk 8]
```diff
@@ -28,29 +28,49 @@ Functions to perform should be stored in the following json format:
...
     ...etc...
   ]
 }
```
Comment (hgrehm) [inline README.md:76]: Specifications below this block should illustrate the new requirements. For example, `function: period-delimited path to a callable function followed by function name, separated by a colon` and `inputs: list of arguments to the given function, which the user must confirm are the correct type`

File: scheduler.py [hunk 9]
```diff
@@ -196,18 +233,9 @@ def main():
...
-
-            ensemble = active_ensembles["ensemble_list"][next_ensemble]
-
-            # perform ensemble functions
-            for (auto f : ensemble)
-            	std::invoke(f, inputs) # Wesley
-            	# TODO: allow for function arguments via e["inputs"]
-        		# should just be comma-separated list of inputs
+            print("performing ensemble functions: ")
+            perform_ensemble_functions(0) # TODO Iterate through the ensemble functions
```
Comment (hgrehm) [inline scheduler.py]: We should be performing the current ensemble specified in `active_ensembles.json`. Specifically, this should be in the `next_ensemble` variable.

File: scheduler.py [hunk 10]
```diff
@@ -105,12 +110,44 @@ def setup(filename: str = "dummy_ensembles.json"):
     ensemble_ofile.close()
 
 
+
 def someFunc():
     ''''
     simple printing function for use in testing
     '''
     print("someFunc called!")
```
Comment (hgrehm) [inline scheduler.py]: Can delete this if it's not being used for testing

**Pull request merged**